### PR TITLE
Implement client Lock idempotence 

### PR DESCRIPTION
**Pull request by ihsandemir**
This PR adds the client Lock idempotence in accordance with the work performed at the member side locks in this PR: https://github.com/hazelcast/hazelcast/pull/5602 and this issue: https://github.com/hazelcast/hazelcast/issues/8359 

The following changes are performed in this PR:

1. Added lockReferenceId support for Lock, Map and Multimap.

2. Updated the client binary compatibility and server compatibility tests so that they can handle older protocol versions (see corresponding generator PR: https://github.com/hazelcast/hazelcast-client-protocol/pull/16 ).

3. Changed the client protocol version to SNAPSHOT version.

### CODE REVIEW

**Comment by mustafaiman** [comment]
👍

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java [hunk 1]
```diff
@@ -179,6 +182,8 @@
     private final SerializationService serializationService;
     private final ClientICacheManager hazelcastCacheManager;
 
+    private ClientLockReferenceCounter lockReferenceCounter;
```
Comment (mdogan) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: why not `final`?

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java [hunk 2]
```diff
@@ -435,12 +442,19 @@ public String getName() {
 
     @Override
     public <K, V> IMap<K, V> getMap(String name) {
-        return getDistributedObject(MapService.SERVICE_NAME, name);
+        DistributedObject map = getDistributedObject(MapService.SERVICE_NAME, name);
+        ClientMapProxy<K, V> mapProxy = (ClientMapProxy<K, V>)map;
+        mapProxy.setReferenceCounter(lockReferenceCounter);
```
Comment (mdogan) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: Isn't it better to set referenceCounter in `ClientProxy.onInitialize() ` ?

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/ClientLockReferenceCounter.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.impl;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public final class ClientLockReferenceCounter {
```
Comment (mdogan) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/ClientLockReferenceCounter.java]: `...ReferenceCounter` name looks misleading. It doesn't count references, it's just creating/generating reference-id. Something like generator, factory...

Comment (mdogan) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/ClientLockReferenceCounter.java]: Can you also add a few lines of comment about purpose of this class?

**Comment by ihsandemir** [comment]
verify

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java (hunk 1 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: OK, changed it.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java (hunk 2 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: OK, changed. It looks much better.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/ClientLockReferenceCounter.java (hunk 3 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/ClientLockReferenceCounter.java]: OK, done. Changed to ClientLockReferenceIdGenerator.

**Comment by mdogan** [comment]
Can you also squash the commits? Currently some of them seem redundant.

**Comment by mdogan** [comment]
👍

**Comment by ihsandemir** [comment]
verify

**Comment by ihsandemir** [comment]
verify

**Pull request merged**